diesel/src/sqlite/types/json.rs

```diff
@@ -0,0 +1,164 @@
+        serde_json::from_str(value.read_text()).map_err(|_| "Invalid Json".into())
+    }
+}
+
+#[cfg(all(feature = "sqlite", feature = "serde_json"))]
+impl ToSql<sql_types::Json, Sqlite> for serde_json::Value {
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
+        out.set_value(serde_json::to_string(self)?);
+        Ok(IsNull::No)
+    }
+}
+
+#[cfg(all(feature = "sqlite", feature = "serde_json"))]
+impl FromSql<sql_types::Jsonb, Sqlite> for serde_json::Value {
+    fn from_sql(value: SqliteValue<'_, '_, '_>) -> deserialize::Result<Self> {
+        let bytes = value.read_blob();
+        if bytes[0] != 1 {
+            return Err("Unsupported JSONB encoding version".into());
+        }
+        serde_json::from_slice(&bytes[1..]).map_err(|_| "Invalid Json".into())
+    }
+}
+
+#[cfg(all(feature = "sqlite", feature = "serde_json"))]
+impl ToSql<sql_types::Jsonb, Sqlite> for serde_json::Value {
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
+        out.set_value(serde_json::to_string(self)?.into_bytes());
+        Ok(IsNull::No)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::deserialize::FromSql;
+    use crate::dsl::sql;
+    use crate::prelude::*;
+    use crate::select;
+    use crate::serialize::{Output, ToSql};
+    use crate::sql_types;
+    use crate::sql_types::{Json, Text};
+    use crate::sqlite::connection::SqliteBindCollector;
+    use crate::sqlite::Sqlite;
+    use crate::sqlite::SqliteBindValue;
+    use crate::sqlite::SqliteValue;
+    use crate::test_helpers::connection;
+    use serde_json::json;
+
+    // #[test]
+    // fn json_to_sql() {
+    //     let buffer = SqliteBindValue::from(0i32);
+    //     let mut out = Output::<'_, 'static, Sqlite>::test(buffer);
+    //     let test_json = serde_json::Value::Bool(true);
+    //     ToSql::<sql_types::Json, Sqlite>::to_sql(&test_json, &mut out).unwrap();
+    //     assert_eq!(buffer.inner, SqliteBindValue::from(1i32).inner);
+    // }
+
+    // #[test]
+    // fn json_to_sql() {
+    //     crate::table! {
+    //         #[allow(unused_parens)]
+    //         test_insert_json_into_table_as_text(id) {
+    //             id -> Integer,
+    //             json -> Text,
+    //         }
+    //     }
+    //     let conn = &mut connection();
+    //     crate::sql_query(
+    //         "CREATE TABLE test_insert_json_into_table_as_text(id INTEGER PRIMARY KEY, json TEXT);",
+    //     )
+    //     .execute(conn)
+    //     .unwrap();
+
+    //     let value = json!(true);
+
+    //     crate::insert_into(test_insert_json_into_table_as_text::table)
+    //         .values((
+    //             test_insert_json_into_table_as_text::id.eq(1),
+    //             test_insert_json_into_table_as_text::json.eq(value),
+    //         ))
+    //         .execute(conn)
+    //         .unwrap();
+    // }
+
+    #[test]
+    fn some_json_from_sql() {
```

Comment:
@weiznich I am having a lot of difficulties writing the tests. I tried to mimic what's done for Pg or Sqlite/Chrono, without success.

I can't find the SQLite equivalent for:

```rust
#[test]
    fn json_to_sql() {
        let mut buffer = Vec::new();
        let mut bytes = Output::test(ByteWrapper(&mut buffer));
        let test_json = serde_json::Value::Bool(true);
        ToSql::<sql_types::Json, Pg>::to_sql(&test_json, &mut bytes).unwrap();
        assert_eq!(buffer, b"true");
    }
```

The SQLite interface for `Output` is completely different.

Any help would be greatly appreciated.

Reply:
@weiznich thank you! Here is what I tried:

```rust
#[test]
    fn json_to_sql() {
        let conn = &mut connection();
        let value = json!(true);
        let res = diesel::select(value.into_sql::<Jsonb>().eq(&sql("true")))
            .get_result::<bool>(conn)
            .unwrap();
        assert!(res);
    }
```

But it does not work:

```
error[E0277]: the trait bound `Value: expression::Expression` is not satisfied
   --> diesel/src/sqlite/types/json.rs:521:40
    |
521 |         let res = diesel::select(value.into_sql::<Jsonb>().eq(sql("true")))
    |                                        ^^^^^^^^ the trait `expression::Expression` is not implemented for `Value`, which is required by `Value: expression::AsExpression<Jsonb>`
    |
    = help: the following other types implement trait `expression::Expression`:
              &'a T
              (T0, T1)
              (T0, T1, T2)
              (T0, T1, T2, T3)
              (T0, T1, T2, T3, T4)
              (T0, T1, T2, T3, T4, T5)
              (T0, T1, T2, T3, T4, T5, T6)
              (T0, T1, T2, T3, T4, T5, T6, T7)
            and 122 others
note: required for `Value` to implement `expression::AsExpression<Jsonb>`
```

Reply:
@weiznich [This](https://github.com/diesel-rs/diesel/pull/4284#discussion_r1775089572) + `use crate::ExpressionMethods;` fixed the build of [that test](https://github.com/diesel-rs/diesel/pull/4284#discussion_r1774135529).

And the following test passes:

```rust
#[test]
    fn json_to_sql() {
        let conn = &mut connection();
        let value = json!(true);
        let res = diesel::select(value.into_sql::<Json>().eq(&sql("json('true')")))
            .get_result::<bool>(conn);
        assert!(res.unwrap());
    }

    #[test]
    fn jsonb_to_sql() {
        let conn = &mut connection();
        let value = json!(true);
        let res = diesel::select(value.into_sql::<Jsonb>().eq(&sql("jsonb('true')")))
            .get_result::<bool>(conn);
        assert!(res.unwrap());
    }
```

I' ll make them more comprehensive and commit them.